leoism/step/walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,124 @@
+   */
+  private TimeRange doesOverlap(TimeRange requestedTime, ArrayList<TimeRange> slotsTaken) {
+    for (TimeRange occupied : slotsTaken) {
+      if (requestedTime.overlaps(occupied)) {
+        // If an overlap occurs, merge them by getting the earliest start time and the latest start
+        // time.
+        int startTime =
+            occupied.start() > requestedTime.start() ? requestedTime.start() : occupied.start();
+        int endTime = occupied.end() > requestedTime.end() ? occupied.end() : requestedTime.end();
+        slotsTaken.remove(occupied);
+        return TimeRange.fromStartEnd(startTime, endTime, false);
+      }
+    }
+    return null;
+  }
+
+  /**
+   * Checks if all required attendees (those that the event is booked under) are the same as the one
+   * making the Meeting Request. Returns false if a required attendee is not present. Otherwise,
+   * returns true.
+   */
+  private boolean areRequiredAttendeesPresent(
+      Collection<String> requiredAttendees, Collection<String> presentAttendees) {
+    for (String attendee : presentAttendees) {
+      if (!requiredAttendees.contains(attendee)) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  /**
+   * Stores all attendees which the event is booked under and stores them inside a Set to avoid
+   * duplicates. Returns a HashSet with all attendees booked under an event.
+   */
+  private HashSet<String> allEventAttendees(Collection<Event> events) {
+    HashSet<String> eventAttendees = new HashSet<String>();
+    for (Event event : events) {
+      eventAttendees.addAll(event.getAttendees());
+    }
+
+    return eventAttendees;
+  }
+
+  /** Returns an ArrayList of TimeRange of all slots that are occupied by events. */
+  private ArrayList<TimeRange> getSlotsTaken(Collection<Event> events, MeetingRequest request) {
+    ArrayList<TimeRange> slotsTaken = new ArrayList<TimeRange>();
+    Collection<String> mandatoryAttendees = request.getAttendees();
+    Collection<String> eventAttendees = allEventAttendees(events);
+
+    for (Event event : events) {
+      TimeRange requestedTime = event.getWhen();
+      TimeRange overlappedCombined = doesOverlap(requestedTime, slotsTaken);
+
+      boolean mandatoryAttendeesPresent =
+          areRequiredAttendeesPresent(eventAttendees, mandatoryAttendees);
+      if (!mandatoryAttendeesPresent) {
+        continue;
+      }
+
+      if (overlappedCombined != null) {
+        slotsTaken.add(overlappedCombined);
+        continue;
+      }
+      slotsTaken.add(requestedTime);
+    }
+
+    return slotsTaken;
+  }
+
+  /**
+   * Checks all available slots to see if the there is sufficient time for the requested meeting
+   * duration. If there is a sufficient time slot, it is added to an ArrayList of TimeRange. When
+   * all slots have been checked, returns an ArrayList of TimeRange with sufficient times.
+   */
+  private ArrayList<TimeRange> sufficientTimeAvailable(
```

Comment:
sufficientTimesAvailable? (ie you're returning Times. As is it sounds like the function is about if any time is available

Reply:
Done, changed the name to getAvailableTimes